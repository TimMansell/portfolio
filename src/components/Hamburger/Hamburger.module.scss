@import 'variables';
@import 'utilities/pxtorem';

.container {
  position: fixed;
  z-index: 999999;
  bottom: 0;
  transform: translate(calc(50vw - 50%), -35%);
}

.hamburger {
  text-indent: -9999px;
  border: 0;
  transition: background 0.3s;
  width: pxtorem(60);
  height: pxtorem(60);
  border-radius: 50%;
  box-shadow: 0 0 10px var(--colour-2);
  background-color: var(--colour-1);

  @media (min-width: $bp-lg) {
    display: none;
  }

  &:focus {
    outline: none;
  }

  &Active {
    background-color: var(--colour-3);
  }
}

.background {
  position: absolute;
  border-radius: 50%;
  transform: scale(0);
  animation: active 0.3s linear;
  animation-fill-mode: forwards;
  background-color: rgba(#000, 0.8);
  width: 100%;
  height: 100%;
  top: 0;
  left: 0;
  z-index: -1;
}

@keyframes active {
  to {
    transform: scale(40);
  }
}

.menu {
  display: block;
  position: absolute;
  left: pxtorem(18);
  right: pxtorem(18);
  height: pxtorem(3);
  background-color: var(--secondary-text-colour);
  transition: background 0 0.3s;
  box-shadow: -1px -1px var(--colour-1-alt);

  &::after,
  &::before {
    position: absolute;
    left: 0;
    width: 100%;
    height: pxtorem(3);
    background-color: var(--secondary-text-colour);
    content: '';
    transition-duration: 0.3s, 0.3s;
    transition-delay: 0.3s, 0;
    box-shadow: -1px -1px var(--colour-1-alt);
  }

  &::before {
    top: pxtorem(-10);
    transition-property: top, transform;
  }

  &::after {
    bottom: pxtorem(-10);
    transition-property: bottom, transform;
  }

  &Active {
    background: none;
    box-shadow: none;

    &::before {
      top: 0;
      transform: rotate(45deg);
    }

    &::after {
      bottom: 0;
      transform: rotate(-45deg);
    }

    &::before,
    &::after {
      transition-delay: 0, 0.3s;
      box-shadow: none;
    }
  }
}
