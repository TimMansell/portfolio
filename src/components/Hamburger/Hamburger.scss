@import 'variables';
@import 'utilities/breakpoints';

$mobile-menu-bg-colour: #fff;

.hamburger {
	position: fixed;
	text-indent: -9999px;
	border: none;
	transition: background 0.3s;
	z-index: 999999;
	left: 50%;
	transform: translate(-50%, -35%);
	width: 60px;
	height: 60px;
	border-radius: 50%;
	box-shadow: 0 0 10px rgba(darken($primary-colour, 30%), .5);
	background-color: $primary-colour;
	bottom: 0;

	@include media-breakpoint-up(lg){
		display: none;
	}

	&:focus {
		outline: none;
	}

	&--active {
		background-color: $tertiary-colour;
	}

	&__menu {
		display: block;
		position: absolute;
		left: 18px;
		right: 18px;
		height: 3px;
		background-color: $mobile-menu-bg-colour;
		transition: background 0 0.3s;
		box-shadow: -1px -1px darken($primary-colour, 5%);

		&:after, &:before {
			position: absolute;
			left: 0;
			width: 100%;
			height: 3px;
			background-color: $mobile-menu-bg-colour;
			content: "";
			transition-duration: 0.3s, 0.3s;
			transition-delay: 0.3s, 0;
			box-shadow: -1px -1px darken($primary-colour, 5%);
		}

		&:before {
			top: -10px;
			transition-property: top, transform;
		}

		&:after {
			bottom: -10px;
			transition-property: bottom, transform;
		}

		&--active {
			background: none;
			box-shadow: none;
	
			&:before {
				top: 0;
				transform: rotate(45deg);
			}
	
			&:after {
				bottom: 0;
				transform: rotate(-45deg);
			}
	
			&:before,
			&:after {
				transition-delay: 0, 0.3s;
				box-shadow: none;
			}
		}
	}
}