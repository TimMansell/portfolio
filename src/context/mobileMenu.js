import React, { useState, createContext } from 'react';
import PropTypes from 'prop-types';

export const MenuContext = createContext();

export const MenuContextProvider = ({ children }) => {
  const [isMobileMenu, setIsMobileMenu] = useState(false);

  return (
    <MenuContext.Provider value={[isMobileMenu, setIsMobileMenu]}>
      {children}
    </MenuContext.Provider>
  );
};

MenuContextProvider.propTypes = {
  children: PropTypes.node.isRequired
};
