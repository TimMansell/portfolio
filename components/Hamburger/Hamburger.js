import React, {useContext, useState} from "react";
import {MenuContext} from "../../context/mobileMenu";
import classnames from "classnames";

import styles from "./Hamburger.module.scss";

export const Hamburger = () => {
  const [isMobileMenu, setIsMobileMenu] = useContext(MenuContext);
  const [isClicked, setIsClicked] = useState(false);

  const toggleHamburger = () => {
    setIsMobileMenu(!isMobileMenu);
    setIsClicked(true);
  };

  const buttonClasses = classnames(styles.hamburger, {
    [styles.hamburgerIsActive]: isMobileMenu,
  });

  const hamburgerClasses = classnames(styles.menu, {
    [styles.menuIsActive]: isMobileMenu,
  });

  const backgroundClasses = classnames(styles.background, {
    [styles.backgroundIsOpen]: isMobileMenu,
    [styles.backgroundIsClosed]: !isMobileMenu && isClicked,
  });

  return (
    <div className={styles.container}>
      <button
        type="button"
        data-e2e="hambuger"
        className={buttonClasses}
        onClick={() => toggleHamburger()}
      >
        <span className={hamburgerClasses}>toggle menu</span>
      </button>
      <div className={backgroundClasses}></div>
    </div>
  );
};

export default Hamburger;
