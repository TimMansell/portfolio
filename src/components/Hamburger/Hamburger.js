import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import classnames from 'classnames';

import * as actions from 'actions';

import './Hamburger.scss';

export class Hamburger extends React.Component {
  toggleHamburger = () => { 
    const { isMobileMenu } = this.props;

    this.props.actions.setMobileMenu(!isMobileMenu);
  }

  render() {
    const { isMobileMenu } = this.props;

    const buttonClasses = classnames('hamburger', {
      'hamburger--active': isMobileMenu
    });

    const hamburgerClasses = classnames('hamburger__menu', {
      'hamburger__menu--active': isMobileMenu
    });
    
    return <button type="button" className={buttonClasses} onClick={this.toggleHamburger}>
      <span className={hamburgerClasses}>toggle menu</span>
    </button>;
  }
}

function mapStateToProps(state, ownProps) {
  return {
    isMobileMenu: state.isMobileMenu
  };
}

const mapDispatchToProps = (dispatch) => {
  return {
    actions: bindActionCreators(actions, dispatch)
  };
};

Hamburger.propTypes = {
	isMobileMenu: PropTypes.bool
};

export default connect(mapStateToProps, mapDispatchToProps)(Hamburger);
