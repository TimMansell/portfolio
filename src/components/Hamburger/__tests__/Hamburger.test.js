import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom';

import Hamburger from '../Hamburger';
import { MenuContext } from '../../../context/mobileMenu';

const MockComponent = () => (
  <MenuContext.Provider value={[false, jest.fn()]}>
    <Hamburger />
  </MenuContext.Provider>
);

describe('Hamburger', () => {
  it('should render my component', () => {
    render(<MockComponent />);
  });

  it('should match snapshot', () => {
    const { asFragment } = render(<MockComponent />);

    expect(asFragment()).toMatchSnapshot();
  });
});
