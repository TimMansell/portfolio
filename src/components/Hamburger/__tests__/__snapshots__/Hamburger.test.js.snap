// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hamburger should match snapshot 1`] = `
<button
  className="hamburger"
  data-e2e="hambuger"
  onClick={[Function]}
  type="button"
>
  <span
    className="hamburger__menu"
  >
    toggle menu
  </span>
</button>
`;
